src/promptflow/promptflow/_core/openai_injector.py

```diff
@@ -81,63 +119,106 @@ def get_aoai_telemetry_headers() -> dict:
 
 
-def inject(f):
-    wrapper_fun = inject_operation_headers((inject_function(["api_key", "headers"])(f)))
+def inject_async(f):
+    wrapper_fun = inject_operation_headers((inject_function_async(["api_key", "headers", "extra_headers"])(f)))
     wrapper_fun._original = f
     return wrapper_fun
 
 
-def available_openai_apis():
+def inject_sync(f):
+    wrapper_fun = inject_operation_headers((inject_function_sync(["api_key", "headers", "extra_headers"])(f)))
+    wrapper_fun._original = f
+    return wrapper_fun
+
+
+def available_openai_apis_and_injectors():
+    """
+    Generates a sequence of tuples containing OpenAI API classes, method names, and
+    corresponding injector functions based on whether the legacy OpenAI interface is used.
+
+    This function handles the discrepancy reported in https://github.com/openai/openai-python/issues/996,
+    where async interfaces were not recognized as coroutines. It ensures that decorators
+    are applied correctly to both synchronous and asynchronous methods.
+
+    Yields:
+        Tuples of (api_class, method_name, injector_function)
+    """
+    # Define API tuples for legacy and new OpenAI interfaces
     if IS_LEGACY_OPENAI:
-        for api in ("Completion", "ChatCompletion", "Embedding"):
-            try:
-                openai_api = getattr(openai, api)
-                if hasattr(openai_api, "create"):
-                    yield openai_api
-            except AttributeError:
-                # This is expected for older versions of openai or unsupported APIs.
-                # E.g. ChatCompletion API was introduced in 2023 and requires openai>=0.27.0 to work.
-                # Older versions of openai do not have this API and will raise an AttributeError if we try to use it.
-                pass
+        sync_apis = (
+            (openai, "Completion", "create", inject_sync),
+            (openai, "ChatCompletion", "create", inject_sync),
+            (openai, "Embedding", "create", inject_sync),
+        )
+
+        async_apis = (
+            (openai, "Completion", "acreate", inject_async),
+            (openai, "ChatCompletion", "acreate", inject_async),
+            (openai, "Embedding", "acreate", inject_async),
+        )
     else:
-        for api in ("Completions", "Chat", "Embeddings"):
+        sync_apis = (
+            (openai.resources.chat, "Completions", "create", inject_sync),
+            (openai.resources, "Completions", "create", inject_sync),
+            (openai.resources, "Embeddings", "create", inject_sync),
+        )
+
+        async_apis = (
+            (openai.resources.chat, "AsyncCompletions", "create", inject_async),
+            (openai.resources, "AsyncCompletions", "create", inject_async),
+            (openai.resources, "AsyncEmbeddings", "create", inject_async),
+        )
+
+    # Helper function to check if the API methods exist and yield the valid tuples
+    def check_apis(apis):
+        for module, clz, method, injector in apis:
```

Comment:
`clz` is usually used to indicate an instance of a class. So do the `method`.
I'd prefer to use `class_name`, `method_name` here to indicate they are names (not instances).
On the other hand, `method` is an instance, so it is a good name here.
```suggestion
        for module, class_name, method_name, injector in apis:
```